setup_configure.py

```diff
@@ -43,131 +44,81 @@ def stash_config(dct):
+        h5_inc, h5_lib, h5_macros = cls._find_hdf5_compiler_settings(mpi)
 
-        Options can come from either command line options or environment
-        variables (but specifying the same option in both is an error).
-        Options not specified will be loaded from the previous configuration,
-        so they are 'sticky' (except hdf5-version).
+        h5_version_s = os.environ.get('HDF5_VERSION')
+        if h5_version_s:
+            h5_version = validate_version(h5_version_s)
+        else:
+            h5_version = autodetect_version(h5_lib)
 
-        When options change, the rebuild_required attribute is set, and
-        may only be reset by calling reset_rebuild().  The custom build_ext
-        command does this.
-    """
+        return cls(h5_inc, h5_lib, h5_macros, h5_version, mpi)
+
+    @staticmethod
+    def _find_hdf5_compiler_settings(mpi=False):
+        """Get compiler settings from environment or pkgconfig.
 
-    description = "Configure h5py build options"
-
-    user_options = [('hdf5=', 'h', 'Custom path prefix to HDF5'),
-                    ('hdf5-version=', '5', 'HDF5 version "X.Y.Z"'),
-                    ('hdf5-includedir=', 'i', 'path to HDF5 headers'),
-                    ('hdf5-libdir=', 'l', 'path to HDF5 library'),
-                    ('hdf5-pkgconfig-name=', 'p', 'name of HDF5 pkgconfig file'),
-                    ('mpi', 'm', 'Enable MPI building'),
-                    ('reset', 'r', 'Reset config options') ]
-
-    def initialize_options(self):
-        self.hdf5 = None
-        self.hdf5_version = None
-        self.hdf5_includedir = None
-        self.hdf5_libdir = None
-        self.hdf5_pkgconfig_name = None
-        self.mpi = None
-        self.reset = None
-
-    def finalize_options(self):
-        # Merge environment options with command-line
-        for setting, env_val in get_env_options().items():
-            if env_val is not None:
-                if getattr(self, setting) is not None:
-                    raise ValueError(
-                        f"Provide {setting} in command line or environment "
-                        f"variable, not both."
-                    )
-                setattr(self, setting, env_val)
+        Returns (include_dirs, lib_dirs, define_macros)
+        """
+        hdf5 = os.environ.get('HDF5_DIR')
+        hdf5_includedir = os.environ.get('HDF5_INCLUDEDIR')
+        hdf5_libdir = os.environ.get('HDF5_LIBDIR')
+        hdf5_pkgconfig_name = os.environ.get('HDF5_PKGCONFIG_NAME')
 
         if sum([
-            bool(self.hdf5_includedir or self.hdf5_libdir),
-            bool(self.hdf5),
-            bool(self.hdf5_pkgconfig_name)
+            bool(hdf5_includedir or hdf5_libdir),
+            bool(hdf5),
+            bool(hdf5_pkgconfig_name)
         ]) > 1:
             raise ValueError(
                 "Specify only one of: HDF5 lib/include dirs, HDF5 prefix dir, "
                 "or HDF5 pkgconfig name"
             )
```

Comment:
Should we get rid of this check? It's meant to avoid confusion where a person thinks *this* option applies but the code would actually use *that* option, by rejecting any ambiguity. That might be less likely with fewer ways to specify the same things.

OTOH, "In the face of ambiguity, refuse the temptation to guess"  is still a good principle.

Reply:
I think we'd want to warn at a minimum, even if we don't error.

Reply:
I'll leave it as an error for now. I think it's easy to miss build warnings, especially when there's lots of other output.